src/main/java/com/google/sps/servlets/AddSaleCardServlet.java

```diff
@@ -0,0 +1,173 @@
+      String endTime = HttpServletUtils.getParameter(request, "endTime", "");
+      String description = HttpServletUtils.getParameter(request, "description", "");
+      String geoHash = HttpServletUtils.getParameter(request, "geoHash", "");
+      float radius = 0f;
+      float latitude = 0f;
+      float longitude = 0f;
+      GeoPt vendorLocation = new GeoPt(0f, 0f);
+      
+      // Check values are valid format
+      try {
+        LocalTime start = LocalTime.parse(startTime);
+        LocalTime end = LocalTime.parse(endTime);
+        radius = Float.parseFloat(HttpServletUtils.getParameter(request, "radius", "1000"));
+        latitude = Float.parseFloat(HttpServletUtils.getParameter(request, "lat", "360"));
+        longitude = Float.parseFloat(HttpServletUtils.getParameter(request, "long", "360"));
+        vendorLocation = new GeoPt(latitude, longitude);
+      } catch (DateTimeParseException e) {
+        System.out.println("Bad format to parse: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (NumberFormatException e) {
+        System.out.println("The string is not a parsable float: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (IllegalArgumentException e) {
+        System.out.println("Latitude and/or longitude outside legal range: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      // Get the BlobKey of the image that the user uploaded to Blobstore
+      BlobKey imageBlobKey = getUploadedFileBlobKey(request, "imageFile");
+      String altText = HttpServletUtils.getParameter(request, "altText", "");
+
+      // Check values are not empty or outside range
+      if (businessName.isEmpty() || description.isEmpty() || geoHash.isEmpty() || altText.isEmpty() || 
+          imageBlobKey == null || radius < MIN_DISTANCE || radius > MAX_DISTANCE ) {
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+      // TODO: Create vendor key using the Auth ID
+      // Key vendorKey = KeyFactory.createKey("Vendor", (String) user.authID);
+
+      // Generate Key when datastore put
+      Entity saleCard = new Entity("SaleCard"/*, vendorKey*/);
+      datastore.put(saleCard);      
+
+      // TODO: Remove from datastore existing saleCard, picture and location data
+
+      Entity picture = new Entity("Picture", saleCard.getKey());
+      picture.setProperty("blobKey", imageBlobKey);
+      picture.setProperty("altText", altText);
+      datastore.put(picture);
+
+      Entity locationData = new Entity("LocationData", saleCard.getKey());
+      locationData.setProperty("salePoint", vendorLocation);
+      locationData.setProperty("geoHash", geoHash);
+      locationData.setProperty("radius", radius);
+      datastore.put(locationData);
```

Comment:
I was thinking, since a Vendor only have one SaleCard, and a SaleCard only have one Picture and one LocationData.

Instead of doing a query to delete previous SaleCards, Pictures and Locations, we can just set an ID/Name when creating the Entity, e.g. 'thisIsMyEntity', this way when `datastore.put(...)` is called, it updates the same Entity with new values.

Something like this:
```
final String UNIQUE = "thisEntityBelongsTo";
. . .
Entity saleCard = new Entity("SaleCard", UNIQUE, vendorKey);
. . .
Entity picture = new Entity("Picture", UNIQUE, saleCard.getKey());
. . .
```
Each entity will still be unique thanks to the Ancestor Paths.
So each key would look something like 
```
OneSaleCard Key: VendorOneKey + "thisEntityBelongsTo"
AnotherSaleCard Key: VendorTwoKey + "thisEntityBelongsTo"
. . .
OneLocationData Key: OneSaleCardKey + "thisEntityBelongsTo"
AnotherLocationData Key: AnotherSaleCardKey + "thisEntityBelongsTo"
```
Long comment (hope I explained it well) 😅 , but what do you think?

Reply:
I like the idea. Just one question, weren't we considering that each sale card would have a `List<Pictures>`? Can this idea still be done with this implementation?

Reply:
Yep, I decided to keep it as just one picture for the moment. And yes, we can still do this implementation with a List, I am still clearing up details in my mind, but I think it should work just by using the Vendor Entity.

Reply:
Great, thanks for the explanation.